homeassistant/components/mqtt/client.py

```diff
@@ -808,6 +813,13 @@ def _mqtt_handle_message(self, msg: mqtt.MQTTMessage) -> None:
         subscriptions = self._matching_subscriptions(msg.topic)
 
         for subscription in subscriptions:
+            init_status = self._retained_init.setdefault(subscription, set())
+            # skip already initialized subscriptions
+            if msg.retain and msg.topic in init_status:
+                continue
+            # remember the subscription had an initial payload
+            self._retained_init[subscription].add(msg.topic)
```

Comment:
SO current logic is:

IF we have multiple subscriptions for the same topic (or more correctly, subsequent subscriptions to the same topic)
IF a message comes in (RETAIN=0) AFTER we have created the Subscription class, but BEFORE we actually send a new SUBSCRIBE to the MQTT server (ie. we are in the debounce period to not flood the MQTT server with SUBSCRIBE calls)
THEN send the message to the Subscription callback, even though, technically, the MQTT server never got the SUBSCRIBE for it yet (it's still in the pending list).
THEN the SUBSCRIBE message will still be sent to the MQTT server, even though that Subscription class has has already processed messages.

Is that about right?  If so, stop calling it retained_init.  At this point, the retained_init has nothing to do with any RETAIN=1 message.  It just means 'I processed at least once message on this Subscription class' (RETAIN = 0 or 1).

Indeed, you could just remember if you have gotten a RETAIN=1 message for the TOPIC (not subscription) and if you have, ignore all future RETAIN=1 messages for that topic, and remember the last message on that topic and just re-send that to any new Subscription class you create.  Which is what I originally did.  It would be functionally equivalent here (without the need to wait for a new message to come in).

Though I was told to re-write it because I wasn't treating RETAIN=1 messages correctly.  But neither is this PR.

Reply:
Additionally, a RETAIN=1 message can never be stale.  The MQTT server sends these messages out serially, and we process them serially.  So AT WORST, the RETAIN=1 message would be a duplicate of the immediately prior RETAIN=0 message.  Though IF you end up in the situation of something sending partial data in RETAIN=0 (which you insist can't happen, but absolutely can), then the RETAIN=1 message contains the most up-to-date information for the topic.

Reply:
1. I'm discussing the PR, and the impact of it.  The PR is the appropriate place to discuss it.
2. I don't know you, I don't know anywhere else you hang out, so I have no idea WHERE else I would contact you to discuss this.
3. From this and other PRs I've seen, you seem to employ the 'sacred cow' ideology to coding (ie. "If I wrote it, it MUST be correct! It's sacred!  I won't make any changes just because someone else said I should!  Because I wrote it, it's sacred.").   Along with a dash of Biased coder ideology (ie. "These other coders are also maintainers, so they MUST know what they're doing, so I can accept their criticism.  But everyone else is wrong, because they can't be right.  Because they're not part of the club.")

Thing about Github, is it's public.  If you are a stubborn coder, or a biased coder, or one who needs to have any code faults hammered into you with copious evidence rather than being able to tell you the issue and have you investigate.  It's on public display.

Similarly, if you are a coder who can accept feedback.  Accept their code may have flaws and change it.  And most importantly, if they are right, can explain HOW and WHY they are right, to convince their critics (ie. explain what's going on and why it's correct, rather than just saying "you're wrong" and considering that enough).  That is ALSO on public display.

I've explained the issue.  Multiple times.  You've never refuted it, just complained about my explaining it in too much detail and that's 'too much to read' or 'that's now how MY environment works.'  I doubt a different venue will convince you otherwise.  But if you want to discuss this over a different venue, real-time, by all means, I'll explain it again there.

Reply:
It is disappointing to see your reaction like this. I meant it in a positive way. It is not about me or about what I think is right. The point is that we are tuning in circles and often get of topic. That is fine on it's own, but makes it very hard for other maintainers to follow.

Reply:
about the name `_retained_init` you are right that is also set when receiving an actual value, this is because it is more accurate then any replayed payload. And we do not we want to filterout any retained values from then. But I think the name reflexes the function good enough.